webxr/magicleap/mod.rs

```diff
@@ -0,0 +1,441 @@
+        let draw_fbo = framebuffers[0];
+        let read_fbo = framebuffers[1];
+
+        let in_frame = false;
+        let frame_handle = MLHandle::default();
+        let cameras = MLGraphicsVirtualCameraInfoArray::default();
+
+        let mut device = MagicLeapDevice {
+            gl,
+            graphics_client,
+            head_tracking_sdata,
+            draw_fbo,
+            read_fbo,
+            in_frame,
+            frame_handle,
+            cameras,
+        };
+
+        // Rather annoyingly, in order for the views to be available, we have to
+        // start a frame at this point.
+        device.start_frame()?;
+
+        Ok(device)
+    }
+}
+
+impl MagicLeapDevice {
+    fn start_frame(&mut self) -> Result<(), MLResult> {
+        if !self.in_frame {
+            warn!("Starting frame");
+            let mut params = MLGraphicsFrameParams::default();
+            unsafe { MLGraphicsInitFrameParams(&mut params).ok()? };
+
+            let mut result = unsafe {
+                MLGraphicsBeginFrame(
+                    self.graphics_client,
+                    &params,
+                    &mut self.frame_handle,
+                    &mut self.cameras,
+                )
+            };
+            if result == MLResult::Timeout {
+                warn!("MLGraphicsBeginFrame timeout");
+                let mut sleep = Duration::from_millis(1);
+                let max_sleep = Duration::from_secs(5);
+                // TODO: give up after a while
+                while result == MLResult::Timeout {
+                    sleep = (sleep * 2).min(max_sleep);
+                    warn!(
+                        "MLGraphicsBeginFrame exponential backoff {}ms",
+                        sleep.as_millis()
+                    );
+                    thread::sleep(sleep);
+                    result = unsafe {
+                        MLGraphicsBeginFrame(
+                            self.graphics_client,
+                            &params,
+                            &mut self.frame_handle,
+                            &mut self.cameras,
+                        )
+                    };
+                }
+                warn!("MLGraphicsBeginFrame finished timeout");
+            }
+            result.ok()?;
+
+            let mut snapshot = unsafe { mem::zeroed() };
+            unsafe { MLPerceptionGetSnapshot(&mut snapshot).ok()? };
+
+            let mut transform = MLTransform::default();
+            unsafe {
+                MLSnapshotGetTransform(
+                    snapshot,
+                    &self.head_tracking_sdata.coord_frame_head,
+                    &mut transform,
+                )
+                .ok()?
+            };
+
+            warn!("Started frame");
+            self.in_frame = true;
+        }
+        Ok(())
+    }
+
+    // Approximate the viewer transform by linear interpolation of the two eyes
+    fn lerp_transforms(&self) -> RigidTransform3D<f32, Viewer, Native> {
```

Comment:
ugh but also i see why we need this (Josh needs it too for openxr)

Reply:
Indeed. If we find ourselves doing this often, we may need to rethink the API.